src/mongocxx/test/v1/events.cpp

```diff
@@ -0,0 +1,759 @@
+    }
+}
+
+template <typename MemFn, typename Object, typename Mock>
+void test_server_description(MemFn mem_fn, Mock& mock, Object const& object, identity_type const* identity) {
+    using mongoc_type = typename to_mongoc<Object>::type;
+
+    identity_type const sd1;
+    identity_type sd2;
+
+    auto destroy = libmongoc::server_description_destroy.create_instance();
+    auto new_copy = libmongoc::server_description_new_copy.create_instance();
+
+    destroy
+        ->interpose([&](mongoc_server_description_t* sd) -> void {
+            auto const ptr = static_cast<void const*>(sd);
+            if (ptr && ptr != &sd1 && ptr != &sd2) {
+                FAIL("unexpected mongoc_server_description_t");
+            }
+        })
+        .forever();
+
+    new_copy
+        ->interpose([&](mongoc_server_description_t const* ptr) -> mongoc_server_description_t* {
+            CHECK(static_cast<void const*>(ptr) == &sd1);
+            return reinterpret_cast<mongoc_server_description_t*>(&sd2);
+        })
+        .forever();
+
+    auto fn = mock.create_instance();
+    fn->interpose([&](mongoc_type const* ptr) -> mongoc_server_description_t const* {
+        CHECK(static_cast<void const*>(ptr) == identity);
+        return reinterpret_cast<mongoc_server_description_t const*>(&sd1);
+    });
+
+    server_description const sd = (object.*mem_fn)();
+
+    SECTION("id") {
+        test_uint32(&server_description::id, libmongoc::server_description_id, sd, &sd2);
+    }
+
+    SECTION("round_trip_time") {
+        test_int64(&server_description::round_trip_time, libmongoc::server_description_round_trip_time, sd, &sd2);
+    }
+
+    SECTION("type") {
+        test_string(&server_description::type, libmongoc::server_description_type, sd, &sd2);
+    }
+
+    SECTION("hello") {
+        test_bson(&server_description::hello, libmongoc::server_description_hello_response, sd, &sd2);
+    }
+
+    test_host_and_port(libmongoc::server_description_host, sd, &sd2);
+}
+
+template <typename MemFn, typename Object, typename Mock>
+void test_topology_description(MemFn mem_fn, Mock& mock, Object const& object, identity_type const* identity) {
+    using mongoc_type = typename to_mongoc<Object>::type;
+
+    identity_type const tdid;
+
+    auto fn = mock.create_instance();
+    fn->interpose([&](mongoc_type const* ptr) -> mongoc_topology_description_t const* {
+        CHECK(static_cast<void const*>(ptr) == identity);
+        return reinterpret_cast<mongoc_topology_description_t const*>(&tdid);
+    });
+
+    topology_description const td = (object.*mem_fn)();
+
+    SECTION("type") {
+        auto type = libmongoc::topology_description_type.create_instance();
+
+        char const str = '\0';
+
+        type->interpose([&](mongoc_topology_description_t const* ptr) -> char const* {
+            CHECK(static_cast<void const*>(ptr) == &tdid);
+            return &str;
+        });
+
+        CHECK(td.type() == &str);
```

Comment:
To reuse test helper:

```suggestion
        test_string(&topology_description::type, libmongoc::topology_description_type, td, &tdid);
```